Changes needed to build the project on Linux

mpersano opened the pull request:
Hi, awesome project!

These are the first set of changes I needed to make in order to get the project to build on Linux. I'm also working on making it build with CMake.

Thanks.

chrxh commented:
Hi,
great! I'm looking forward to seeing your changes! Please give me a signal when you are done :)

Alexxxxa commented:
https://github.com/chrxh/alien/pull/24/files

Alexxxxa commented:
https://github.com/chrxh/alien/pull/24/files

Alexxxxa commented:
_[](https://github.com/chrxh/alien/pull/24/files)_![Uploading ![image](https://user-images.githubusercontent.com/83915467/121785152-6f7fce80-cbb8-11eb-9089-e805b2d57fe5.png)121785150...]

r-burns commented:
Oops, should have checked here first, I think I independently made a lot of the same changes. Here's a simple cmake setup if you haven't gotten around to that yet: https://github.com/r-burns/alien/commit/7b81a85d2b626060c6dbc84746d5a76196bc8525

mpersano commented:
@r-burns just saw your branch, looks like we made the same changes.

Did you get it to run? I got it to build (with a CMake script that's not part of this pull request) and the GUI shows up, but I get this error at runtime:

```
CUDA error at CudaSimulation.cu:383 code=999(cudaErrorUnknown) "cudaMemcpyToSymbol( cudaSimulationParameters, &parameters, sizeof(SimulationParameters), 0, cudaMemcpyHostToDevice)"
```

Now I'm wondering if I did something silly or if it's something with my GPU or drivers (I have a GTX 1660 here, so I think it should work).

r-burns commented:
Ha, I'm testing this on the same GPU. I was getting a similar error earlier (although I think it was actually a cuda error 700 triggered by a cudaDeviceSynchronize) but now I'm able to run the simulation without issues, so it may not be a deterministic error. I'm just clicking around to place some dots and hitting play though, I'll try to see if I can load one of the more complex examples.

chrxh commented:
@mpersano: That line is actually the first CUDA command after initialization. It copies the simulation parameters to the constant memory. Maybe this problem is related to 
https://stackoverflow.com/questions/58595291/runtime-error-999-when-trying-to-use-cuda-with-pytorch ?

@mpersano, @r-burns : Sorry, that I've currently no contribution guideline so that we can better coordinate our actions in advance. I should create one.

chrxh commented on source/EngineInterface/SerializerImpl.cpp:
```diff
@@ -150,11 +150,9 @@ namespace boost {
         template<class Archive>
 		inline void serialize(Archive & ar, SimulationParameters& data, const unsigned int /*version*/)
```
This method is not used anymore (I haven't cleaned the code...). Simulation parameters are now serialized to JSON format.

mpersano commented:
@chrxh sorry for the late reply. Now it's working on my setup! There was something stupid in the way I was building the CUDA code. Fixed it after looking at @r-burns's CMake script (thanks!).

I've updated the pull request with all the changes I needed including the CMake scripts.

Ccode-lang commented:
Yay! I'll be able to build this project finally.

chrxh commented:
Good work. I'd like to merge the pull request. On which Linux system did you compile it?

mpersano commented:
Great! I built it on Ubuntu 20.04 (note that Ubuntu's standard repositories don't have CUDA Toolkit 11 or Boost 1.75, so I installed those manually).

Also, I get a boost::archive::archive_exception ("incompatible native format - size of long") when trying to load the example simulations. It looks like the boost.serialization binary format is not portable, unfortunately. :(

chrxh commented:
In the meantime I'd made a change that causes a few useless bytes not to be serialized in the files. Unfortunately, the examples were no longer compatible and I'd to updated them in the repository. Maybe it is related to this?

r-burns commented on CMakeLists.txt:
```diff
@@ -0,0 +1,27 @@
+cmake_minimum_required(VERSION 3.6)
```
CMake 3.18 is required for CMAKE_CUDA_STANDARD 17

mpersano replied:
Thanks!
